marlowe-dashboard-server/src/Marlowe/Run/Webserver/Wallet/CentralizedTestnet/Server.hs

```diff
@@ -0,0 +1,145 @@
+        ExceptT $ getPubKeyHashFromWallet walletId
+
+    pure $ WalletInfo{wiWallet=Pab.Wallet.Wallet (Pab.Wallet.WalletId walletId), wiPubKeyHash = pubKeyHash }
+
+getPubKeyHashFromWallet ::
+    MonadIO m =>
+    MonadReader Env m =>
+    WBE.WalletId ->
+    m (Either ClientError PubKeyHash)
+getPubKeyHashFromWallet walletId = let
+    -- This endpoint is not exposed directly by the WBE, I took this helper from the plutus-pab code.
+    getWalletKey :: WBE.ApiT WBE.WalletId -> WBE.ApiT WBE.Role -> WBE.ApiT WBE.DerivationIndex -> Maybe Bool -> ClientM WBE.ApiVerificationKeyShelley
+    getWalletKey :<|> _ :<|> _ :<|> _ = client (Proxy @("v2" :> WalletKeys))
+
+    makeRequest = callWBE $
+       getWalletKey
+        (WBE.ApiT walletId)
+        -- Role: External, to receive funds
+        (WBE.ApiT WBE.UtxoExternal)
+        -- DerivationIndex: first address
+        (WBE.ApiT (WBE.DerivationIndex 0))
+        -- Return hashed version
+        (Just True)
+
+    transformResponse = (fmap . fmap) (PubKeyHash . BuiltinByteString . fst . getApiVerificationKey)
+  in
+    transformResponse makeRequest
+
+createOrRestoreWallet ::
+    MonadIO m =>
+    MonadReader Env m =>
+    RestorePostData ->
+    WBE.ApiMnemonicT '[15, 18, 21, 24] ->
+    m (Either ClientError WBE.WalletId)
+createOrRestoreWallet postData mnemonic = do
+    let
+        passphrase = Text.unpack $ getPassphrase postData
+        walletName = getWalletName postData
+
+        walletPostData = WBE.WalletOrAccountPostData $ Left $ WBE.WalletPostData
+            Nothing
+            mnemonic
+            Nothing
+            (WBE.ApiT $ WBE.WalletName walletName)
+            (WBE.ApiT $ Passphrase $ fromString passphrase )
+
+    result <- callWBE $ WBE.Api.postWallet WBE.Api.walletClient walletPostData
+    case result of
+        Left err@(FailureResponse _ r) -> do
+            let
+                matchDuplicateWallet :: Regex
+                matchDuplicateWallet = Regex.mkRegex "wallet with the following id: ([a-z0-9]{40})"
+
+                mWalletIdFromErr :: Maybe WBE.WalletId
+                mWalletIdFromErr = do
+                    msg <- decodeError $ responseBody r
+                    matches <- Regex.matchRegex matchDuplicateWallet $ Text.unpack msg
+                    walletIdStr <- headMay matches
+                    rightToMaybe $ fromText $ Text.pack walletIdStr
+            case mWalletIdFromErr of
+                Just walletId -> pure $ Right walletId
+                Nothing       ->  do
+                    putStrLn ("restoreWallet failed" :: Text)
+                    putStrLn $ "Error: " <> (show err :: Text)
+                    pure $ Left err
+        Left err -> do
+            -- FIXME: Define a better logging mechanism
+            putStrLn ("restoreWallet failed" :: Text)
+            putStrLn $ "Error: " <> (show err :: Text)
+            pure $ Left err
+        Right WBE.ApiWallet{WBE.id = WBE.ApiT walletId} -> do
```

Comment:
I modified the pattern match expression from this

```haskell
Right (WBE.ApiWallet (WBE.ApiT walletId) _ _ _ _ _ _ _ _) -> do
```
As I think the new one is clearer.